include/bout/sys/variant.hxx

```diff
@@ -0,0 +1,155 @@
+///
+/// Variant utilities
+///
+/// All in namespace bout::utils
+///    variant
+///    visit
+///    holds_alternative
+///    get
+///
+///    variantEqualTo
+///    variantStaticCastOrThrow
+///    variantToString
+///
+/// Internal implementation in bout::utils::details
+
+#pragma once
+
+#ifndef __VARIANT_HXX__
+#define __VARIANT_HXX__
+
+// std::variant added in C++17
+//#include <variant>
+
+#include "external/mpark/variant.hpp"
+
+#include "utils.hxx"
+
+namespace bout {
+namespace utils {
+
+/// Import variant, visit into bout::utils namespace
+
+// From C++17
+// using std::variant;
+// using std::visit;
+// using std::holds_alternative;
+// using std::get;
+
+using mpark::variant;
+using mpark::visit;
+using mpark::holds_alternative;
+using mpark::get;
+  
+////////////////////////////////////////////////////////////
+// Variant comparison
+
+namespace details {
+
+/// Compare two values.
+///   Different types -> false
+template <typename T, typename U>
+struct CompareTypes {
+  bool operator()(const T& UNUSED(v), const U& UNUSED(t)) { return false; }
+};
+
+/// Compare two values
+///   Same type -> use `==` operator to compare
+template <typename T>
+struct CompareTypes<T, T> {
+  bool operator()(const T& v, const T& t) { return v == t; }
+};
+
+/// A visitor for std::variant which compares
+/// the value stored in the variant with a given value using CompareTypes
+template <typename T>
+struct IsEqual {
+  const T& t;
+  IsEqual(const T& t) : t(t) {}
+
+  template <typename U>
+  bool operator()(const U& u) {
+    return CompareTypes<T, U>()(t, u);
+  }
+};
+} // namespace details
+
+/// Return true only if the given variant \p v
+/// has the same type and value as \p t
+///
+/// Note: Handles the case that \p t is not of a type
+/// which \v can hold.
+template <typename Variant, typename T>
+bool variantEqualTo(const Variant& v, const T& t) {
+  return visit(details::IsEqual<T>(t), v);
+}
+
+////////////////////////////////////////////////////////////
+// Variant casting
+
+namespace details {
+
+/// Helper class for casting between two types
+/// The general case is that casting can't be done
+///
+/// Note: This must be a class/struct because partial
+/// specialisation is not allowed for functions
+template <typename Target, typename Source, bool>
+struct _StaticCastOrThrow {
```

Comment:
I don't think we're allowed to name things with a single underscore followed by an uppercase letter.

Given this is in `bout::details::`, maybe just drop the underscore?